File: ruby/lib/retry.rb
class Retry
  def self.do(count, backoffSeconds)
    laster = nil
    (1...count).each do |x|
      begin
        res = yield
        return res
      rescue => e
        laster = e
        puts laster
        sleep backoffSeconds * x
      end
    end
  
    raise laster
  end
end

File: js/lib/retry.js
const sleep = require('./sleep');
// if fn throws an error, try again in a bit
// if fn throws an error for too long, return (null, er)
// if fn succeeds, return (ret, null)
const retry = async (fn, retrycount = 10, retrybackoff = 1000) => {
  return new Promise(async (resolve, reject) => {
    let count = 0;
    for (;;) {
      try {
        resolve(await fn());
        return; // separate lines, so that resolve the await fn will throw
      } catch (er) {
        if(process.env.DEBUG) {
          console.log('error returned, retrying', er);
        }
        count += 1;
        if (count >= retrycount) {
          return reject(er);
        }
        await sleep(retrybackoff*count);
      }
    }
  });
}

module.exports= retry;